File: src/project/detect.ts
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';

export type ProjectType =
  | 'node'
  | 'python'
  | 'java'
  | 'c-cpp'
  | 'go'
  | 'rust'
  | 'ruby'
  | 'php'
  | 'other'
  | 'unknown';

export type PackageManager =
  | 'npm'
  | 'yarn'
  | 'pnpm'
  | 'bun'
  | 'poetry'
  | 'uv'
  | 'pip'
  | 'maven'
  | 'gradle'
  | 'cargo'
  | 'go'
  | 'none';

export interface ProjectInfo {
  root: string;
  type: ProjectType;
  /** e.g. 'next', 'react', 'vue', 'express', 'nest' — detected from dependencies. */
  framework: string | null;
  packageManager: PackageManager;
  /** Detected languages based on config/entry files. */
  languages: string[];
  configFiles: string[];
  isGitRepo: boolean;
  /** Recommended test/build command hint (may be null). */
  testCommand: string | null;
  buildCommand: string | null;
}

function exists(p: string): boolean {
  try {
    return existsSync(p);
  } catch {
    return false;
  }
}

function readJson(p: string): Record<string, unknown> | null {
  try {
    return JSON.parse(readFileSync(p, 'utf8')) as Record<string, unknown>;
  } catch {
    return null;
  }
}

const FRAMEWORK_HINTS: Array<[string[], string]> = [
  [['next'], 'next'],
  [['nuxt'], 'nuxt'],
  [['react'], 'react'],
  [['vue'], 'vue'],
  [['svelte'], 'svelte'],
  [['@nestjs/core'], 'nest'],
  [['express'], 'express'],
  [['fastify'], 'fastify'],
  [['astro'], 'astro'],
  [['remix'], 'remix'],
  [['gatsby'], 'gatsby'],
];

export function detectProject(root: string): ProjectInfo {
  const configFiles: string[] = [];
  const languages: string[] = [];
  let type: ProjectType = 'unknown';
  let packageManager: PackageManager = 'none';
  let framework: string | null = null;
  let testCommand: string | null = null;
  let buildCommand: string | null = null;

  const packageJson = join(root, 'package.json');
  const hasPkg = exists(packageJson);

  if (hasPkg) {
    type = 'node';
    packageManager = 'npm';
    const pkg = readJson(packageJson) ?? {};
    const scripts = (pkg.scripts ?? {}) as Record<string, string>;
    const deps = { ...((pkg.dependencies ?? {}) as Record<string, string>), ...((pkg.devDependencies ?? {}) as Record<string, string>) };
    for (const [keys, name] of FRAMEWORK_HINTS) {
      if (keys.some((k) => k in deps)) {
        framework = name;
        break;
      }
    }
    if (scripts.test) testCommand = 'npm test';
    if (scripts.build) buildCommand = 'npm run build';
    if (exists(join(root, 'tsconfig.json'))) languages.push('typescript');
    if (exists(join(root, 'jsconfig.json'))) languages.push('javascript');
    if (!languages.length) languages.push('javascript');

    if (exists(join(root, 'pnpm-lock.yaml'))) packageManager = 'pnpm';
    else if (exists(join(root, 'yarn.lock'))) packageManager = 'yarn';
    else if (exists(join(root, 'bun.lockb')) || exists(join(root, 'bun.lock'))) packageManager = 'bun';
  } else if (exists(join(root, 'pyproject.toml'))) {
    type = 'python';
    packageManager = 'uv';
    if (!exists(join(root, 'uv.lock')) && exists(join(root, 'poetry.lock'))) packageManager = 'poetry';
    testCommand = 'python -m pytest';
    languages.push('python');
  } else if (exists(join(root, 'requirements.txt')) || exists(join(root, 'setup.py'))) {
    type = 'python';
    packageManager = 'pip';
    testCommand = 'python -m pytest';
    languages.push('python');
  } else if (exists(join(root, 'pom.xml'))) {
    type = 'java';
    packageManager = 'maven';
    testCommand = 'mvn test';
    languages.push('java');
  } else if (exists(join(root, 'build.gradle')) || exists(join(root, 'build.gradle.kts'))) {
    type = 'java';
    packageManager = 'gradle';
    testCommand = 'gradle test';
    languages.push('java');
  } else if (exists(join(root, 'go.mod'))) {
    type = 'go';
    packageManager = 'go';
    testCommand = 'go test ./...';
    languages.push('go');
  } else if (exists(join(root, 'Cargo.toml'))) {
    type = 'rust';
    packageManager = 'cargo';
    testCommand = 'cargo test';
    languages.push('rust');
  } else if (exists(join(root, 'CMakeLists.txt'))) {
    type = 'c-cpp';
    languages.push('c', 'c++');
  } else if (exists(join(root, 'Makefile')) || exists(join(root, 'makefile'))) {
    type = 'c-cpp';
    languages.push('c', 'c++');
  } else if (exists(join(root, 'Gemfile'))) {
    type = 'ruby';
    languages.push('ruby');
  } else if (exists(join(root, 'composer.json'))) {
    type = 'php';
    languages.push('php');
  } else if (exists(join(root, 'index.html')) || exists(join(root, 'src'))) {
    type = 'other';
  }

  const allConfigs = [
    'package.json', 'tsconfig.json', 'jsconfig.json', 'vite.config.ts', 'vite.config.js',
    'next.config.mjs', 'next.config.js', 'webpack.config.js', 'eslint.config.js', '.eslintrc.json',
    'jest.config.js', 'vitest.config.ts', 'prettier.config.js', '.prettierrc', 'tailwind.config.js',
    'pyproject.toml', 'requirements.txt', 'setup.py', 'pom.xml', 'build.gradle', 'go.mod', 'Cargo.toml',
    'Gemfile', 'composer.json', '.github/workflows/ci.yml', 'Dockerfile', 'docker-compose.yml',
  ];
  for (const c of allConfigs) {
    if (exists(join(root, c))) configFiles.push(c);
  }

  // Common config files detection for node projects
  if (type === 'node') {
    if (exists(join(root, 'vite.config.ts')) || exists(join(root, 'vite.config.js'))) {
      if (!framework) framework = 'vite';
    }
    if (exists(join(root, 'jest.config.js')) || exists(join(root, 'jest.config.ts'))) {
      if (!testCommand) testCommand = 'npm test';
    }
  }

  return {
    root,
    type,
    framework,
    packageManager,
    languages: [...new Set(languages)],
    configFiles,
    isGitRepo: exists(join(root, '.git')),
    testCommand,
    buildCommand,
  };
}

export function projectLabel(info: ProjectInfo): string {
  const bits: string[] = [info.type];
  if (info.framework) bits.push(info.framework);
  return bits.join(' · ');
}

File: grace/project/detect.py
"""Project type/framework detection (port of src/project/detect.ts)."""

import json
import os
from dataclasses import dataclass, field


@dataclass
class ProjectInfo:
    root: str
    type: str = "unknown"  # node | python | java | c-cpp | go | rust | ruby | php | other | unknown
    framework: str | None = None
    packageManager: str = "none"  # npm | yarn | pnpm | bun | poetry | uv | pip | maven | gradle | cargo | go | none
    languages: list[str] = field(default_factory=list)
    configFiles: list[str] = field(default_factory=list)
    isGitRepo: bool = False
    testCommand: str | None = None
    buildCommand: str | None = None


FRAMEWORK_HINTS: list[tuple[list[str], str]] = [
    (["next"], "next"),
    (["nuxt"], "nuxt"),
    (["react"], "react"),
    (["vue"], "vue"),
    (["svelte"], "svelte"),
    (["@nestjs/core"], "nest"),
    (["express"], "express"),
    (["fastify"], "fastify"),
    (["astro"], "astro"),
    (["remix"], "remix"),
    (["gatsby"], "gatsby"),
]


def _exists(p: str) -> bool:
    try:
        return os.path.exists(p)
    except OSError:
        return False


def _read_json(p: str) -> dict | None:
    try:
        with open(p, encoding="utf-8") as f:
            data = json.load(f)
        return data if isinstance(data, dict) else None
    except Exception:
        return None


def detect_project(root: str) -> ProjectInfo:
    config_files: list[str] = []
    languages: list[str] = []
    ptype: str = "unknown"
    package_manager: str = "none"
    framework: str | None = None
    test_command: str | None = None
    build_command: str | None = None

    package_json = os.path.join(root, "package.json")
    has_pkg = _exists(package_json)

    if has_pkg:
        ptype = "node"
        package_manager = "npm"
        pkg = _read_json(package_json) or {}
        scripts = pkg.get("scripts") or {}
        if not isinstance(scripts, dict):
            scripts = {}
        deps = {}
        for section in ("dependencies", "devDependencies"):
            d = pkg.get(section)
            if isinstance(d, dict):
                deps.update(d)
        for keys, name in FRAMEWORK_HINTS:
            if any(k in deps for k in keys):
                framework = name
                break
        if scripts.get("test"):
            test_command = "npm test"
        if scripts.get("build"):
            build_command = "npm run build"
        if _exists(os.path.join(root, "tsconfig.json")):
            languages.append("typescript")
        if _exists(os.path.join(root, "jsconfig.json")):
            languages.append("javascript")
        if not languages:
            languages.append("javascript")

        if _exists(os.path.join(root, "pnpm-lock.yaml")):
            package_manager = "pnpm"
        elif _exists(os.path.join(root, "yarn.lock")):
            package_manager = "yarn"
        elif _exists(os.path.join(root, "bun.lockb")) or _exists(os.path.join(root, "bun.lock")):
            package_manager = "bun"
    elif _exists(os.path.join(root, "pyproject.toml")):
        ptype = "python"
        package_manager = "uv"
        if not _exists(os.path.join(root, "uv.lock")) and _exists(os.path.join(root, "poetry.lock")):
            package_manager = "poetry"
        test_command = "python -m pytest"
        languages.append("python")
    elif _exists(os.path.join(root, "requirements.txt")) or _exists(os.path.join(root, "setup.py")):
        ptype = "python"
        package_manager = "pip"
        test_command = "python -m pytest"
        languages.append("python")
    elif _exists(os.path.join(root, "pom.xml")):
        ptype = "java"
        package_manager = "maven"
        test_command = "mvn test"
        languages.append("java")
    elif _exists(os.path.join(root, "build.gradle")) or _exists(os.path.join(root, "build.gradle.kts")):
        ptype = "java"
        package_manager = "gradle"
        test_command = "gradle test"
        languages.append("java")
    elif _exists(os.path.join(root, "go.mod")):
        ptype = "go"
        package_manager = "go"
        test_command = "go test ./..."
        languages.append("go")
    elif _exists(os.path.join(root, "Cargo.toml")):
        ptype = "rust"
        package_manager = "cargo"
        test_command = "cargo test"
        languages.append("rust")
    elif _exists(os.path.join(root, "CMakeLists.txt")):
        ptype = "c-cpp"
        languages.extend(["c", "c++"])
    elif _exists(os.path.join(root, "Makefile")) or _exists(os.path.join(root, "makefile")):
        ptype = "c-cpp"
        languages.extend(["c", "c++"])
    elif _exists(os.path.join(root, "Gemfile")):
        ptype = "ruby"
        languages.append("ruby")
    elif _exists(os.path.join(root, "composer.json")):
        ptype = "php"
        languages.append("php")
    elif _exists(os.path.join(root, "index.html")) or _exists(os.path.join(root, "src")):
        ptype = "other"

    all_configs = [
        "package.json", "tsconfig.json", "jsconfig.json", "vite.config.ts", "vite.config.js",
        "next.config.mjs", "next.config.js", "webpack.config.js", "eslint.config.js", ".eslintrc.json",
        "jest.config.js", "vitest.config.ts", "prettier.config.js", ".prettierrc", "tailwind.config.js",
        "pyproject.toml", "requirements.txt", "setup.py", "pom.xml", "build.gradle", "go.mod", "Cargo.toml",
        "Gemfile", "composer.json", ".github/workflows/ci.yml", "Dockerfile", "docker-compose.yml",
    ]
    for c in all_configs:
        if _exists(os.path.join(root, c)):
            config_files.append(c)

    if ptype == "node":
        if _exists(os.path.join(root, "vite.config.ts")) or _exists(os.path.join(root, "vite.config.js")):
            if not framework:
                framework = "vite"
        if _exists(os.path.join(root, "jest.config.js")) or _exists(os.path.join(root, "jest.config.ts")):
            if not test_command:
                test_command = "npm test"

    return ProjectInfo(
        root=root,
        type=ptype,
        framework=framework,
        packageManager=package_manager,
        languages=list(dict.fromkeys(languages)),
        configFiles=config_files,
        isGitRepo=_exists(os.path.join(root, ".git")),
        testCommand=test_command,
        buildCommand=build_command,
    )


def project_label(info: ProjectInfo) -> str:
    bits = [info.type]
    if info.framework:
        bits.append(info.framework)
    return " · ".join(bits)
